### PR TITLE
[IPAD-391] Scatter Performance <50K rows

### PR DESCRIPTION
**Pull request by paulnordlund**
The scatterplot had a variety of issues, many stemming from changes that were not accounted for when we connected it to the table. This PR refactors the scatter plot and props that affect it to minimize re-renders, optimize performance. The following bullets summarize the performance bottlenecks (in order of significance) and describe what I've done to address them. 

1. After both box-select and zoom out (double-click), 'transitionZoom' was being run, re-rendering the scatterplot but then informing the table of the new data. The table would adjust, and the scatterplot lifecycle event (that looks for table data changes) would then run the 'transitionZoom' function again, when it didn't need to. I've added a state flag (transitioning) to let that lifecycle event run (after table filter or multi-set analysis) but not after box-select. I've adjusted zoom out to immediately call the parent to change the data, that will trigger the lifecycle event in the scatter plot (in all running once).

2. After scatterplot selection, we were initially just populating the table with brushed data. However, a year or so ago Brett requested that the table data stay in the same order after scatter brushing. To do so, we mapped the ids provided by the brushed scatter plot data, and filtered the table by those ids. This wasn't a noticeable bottleneck with 15K records but for 50K, it's huge. I've commented it out. The good news is that if the table has any sort/filter applied, it will apply that to data the scatter plot gives it. The only drawback is without any active sort/filter, the scatterplot brushed data will dictate the order in the table. If we want, we can turn the calculations on up to a certain threshold (e.g. 15,000 features).

3. Some JavaScript wasn't written in the most efficient way to determine relevant/irrelevant data for binning/circles. After inspecting, it didn't need to account for multi-set filters, that were already included in the current table data. There was also additional flags that could be added to limit calculations.

4. The scatterplot axis options were not being loaded until streaming was finished, and they were being called on an irrelevant update. I've moved all of that html and logic (options, usage, export) into a parent div, and get the axis labels once, when that mounts.

5. The scatterplot component was being mounted and functions were being called before streaming was finished (only to be called again once it did finish). I've restricted this by not even rendering the parent div (4. above ^) until streaming is finished.

6. The 'renderCirclesFilter' function was being passed all of the circles, not just unfiltered. It wasn't visible to the end user because: a) The circles were hidden underneath the unfiltered circles. b) The parent function that took brushed data to render the table filtered them out

7. On double-click (zoom out), there wasn't anything distinguishing it from single click, so both would fire. I've added a 250ms debounce that allows us to distinguish between the two events, and handle separately. Note, I've changed the functionality so that a double-click zoom out doesn't clear highlighted features (which I think makes it more powerful). If users wants to clear highlighted features, they just single click.

### CODE REVIEW

File: src/components/Differential/ScatterPlot.jsx [hunk 1]
```diff
@@ -1300,115 +1192,113 @@ class ScatterPlot extends React.PureComponent {
...
+      //     allFilteredDifferentialTableDataFeatureIds.includes(
+      //       obj[differentialFeatureIdKey],
+      //     )
+      //   ) {
+      //     relevantData.push(obj);
+      //   } else {
+      //     irrelevantData.push(obj);
+      //   }
+      // });
+      [...dataInSelection].forEach(obj => {
```
Comment (paulnordlund) [inline src/components/Differential/ScatterPlot.jsx:1242]: This time-consuming calculation must remain, in order to separate data _within the box selection view_ into relevant (blue) and irrelevant (grey) circles and bins. If you have suggestions for a workaround or how to make it faster, let me know.

**Review by JoeDalen81 — APPROVED** [review]
Awesome job!  The double click is SUPER fast.  I did notice when playing around with GTG590ClariomD -> Cell_Line_Treatment -> H1568_A684_vs_DMSO that the initial render was super fast but if I brush/box select just the bottom half of the screen, where the bins are, it does take a pretty long time to do any subsequent re-renders.  So there might still need to be some additional performance tuning but overall it seems to be better than before.

**Comment by paulnordlund** [comment]
> Member

Yes. As far as I can tell it is purely do to the calculation I've mentioned above, line 1242.  I'm not sure how to make this faster, if you have ideas please share

**Comment by paulnordlund** [comment]
> Awesome job! The double click is SUPER fast. I did notice when playing around with GTG590ClariomD -> Cell_Line_Treatment -> H1568_A684_vs_DMSO that the initial render was super fast but if I brush/box select just the bottom half of the screen, where the bins are, it does take a pretty long time to do any subsequent re-renders. So there might still need to be some additional performance tuning but overall it seems to be better than before.

Yes. As far as I can tell it is purely do to the calculation I've mentioned above, line 1242. I'm not sure how to make this faster, if you have ideas please share

**Review by ternst23 — APPROVED** [review]

**Pull request merged**